src/server/worker/pipeline/transform/worker_v2.go

```diff
@@ -0,0 +1,93 @@
+package transform
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"github.com/pachyderm/pachyderm/src/server/worker/datum"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+)
+
+// WorkerV2 handles a transform pipeline work subtask, then returns.
+// TODO:
+// datum queuing (probably should be handled by datum package).
+// s3 input / gateway stuff (need more information here).
+// spouts.
+// joins.
+// capture datum logs.
+// file download features (empty / lazy files). Need to check over the pipe logic.
+// git inputs.
+// handle custom user set for execution.
+// Taking advantage of symlinks during upload?
+func WorkerV2(driver driver.Driver, logger logs.TaggedLogger, subtask *work.Task, status *Status) (retErr error) {
+	datumSet, err := deserializeDatumSetV2(subtask.Data)
+	if err != nil {
+		return err
+	}
+	return status.withJob(datumSet.JobID, func() error {
+		logger = logger.WithJob(datumSet.JobID)
+		if err := logger.LogStep("datum task", func() error {
+			return handleDatumSetV2(driver, logger, datumSet, status)
+		}); err != nil {
+			return err
+		}
+		subtask.Data, err = serializeDatumSetV2(datumSet)
+		return err
+	})
+}
+
+// TODO: It would probably be better to write the output to temporary file sets and expose an operation through pfs for adding a temporary fileset to a commit.
```

Comment:
Is this still a TODO? Isn't that `CopyFile`

Reply:
Not really, at least as is. We need to preserve the tags. This suggested change is really just to handle the error state where we upload the output, but hit a failure communicating the success of the datum set back to the master. If we don't associate the output with the commit until the master acknowledges the success of the datum set, then we avoid the situation where we upload output for a datum set more than once.